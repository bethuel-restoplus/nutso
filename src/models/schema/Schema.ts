import { ArraySchema } from "./ArraySchema";
import { BooleanSchema } from "./BooleanSchema";
import { DateSchema } from "./DateSchema";
import { NumberSchema } from "./NumberSchema";
import { ObjectSchema } from "./ObjectSchema";
import { StringSchema } from "./StringSchema";
import { TupleSchema } from "./TupleSchema";

// prettier-ignore
export type Schema<T, R = T> = T extends string ? StringSchema<R> :
                        T extends number ? NumberSchema :
                        T extends Date ? DateSchema :
                        T extends boolean ? BooleanSchema :
                        T extends Array<infer E> ? ArraySchema<E, R> : 
                        T extends [infer U, ...unknown[]] ? TupleSchema<U> :
                        T extends Object ? ObjectSchema<T, R> :
                        never;
