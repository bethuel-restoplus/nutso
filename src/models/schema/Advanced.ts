type False = "0";
type True = "1";
type If<C extends True | False, Then, Else> = { "0": Else; "1": Then }[C];

type Diff<T extends string | number | symbol, U extends string> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T];

type X<T> = Diff<keyof T, keyof Object>;

type Is<T, U> = (Record<X<T & U>, False> & Record<any, True>)[Diff<X<T>, X<U>>];

type DeepPartial<T> = {
  [P in keyof T]?: If<Is<Function & T[P], Function>, T[P], DeepPartial<T[P]>>;
};

type Customer = {
  name: string;
  address: {
    line1: string;
    city: string;
    props: { [key: string]: string };
  };
};

type Props = { [key: string]: string };
const props: Props = { 1: "a" };

type C1 = DeepPartial<Customer>;

const c1: C1 = {
  name: "John",
  address: {
    line1: "66 Collins St",
    props: {
      1: "foo"
    }
  }
};

type Foo = {
  type: "foo";
  foo: string;
};

type Bar = {
  type: "bar";
  bar: number;
};

type Baz = Foo | Bar;
const baz: Baz = {
  type: "foo",
  foo: "string"
};
