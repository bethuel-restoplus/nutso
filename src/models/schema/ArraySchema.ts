import { OptionalFlag } from "../OptionalFlag";
import { Schema } from "./Schema";

/**
 * E => Element
 * T => Type (this is the array)
 * P => Parent of the array
 *
 * Look at this type, we need all 3 types
 */
export type ArraySchema<E, T extends E[], P> = {
  type: "array";
  optional?: OptionalFlag<P>;
  minItems?: number;
  maxItems?: number;
  items: Schema<E, T>;
};
