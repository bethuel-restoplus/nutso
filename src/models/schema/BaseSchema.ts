// export type BaseSchema<R> = {
//   optional?: OptionalFlag<R>;
// };
