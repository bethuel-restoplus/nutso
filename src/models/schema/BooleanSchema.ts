import { OptionalFlag } from "../OptionalFlag";

export type BooleanSchema<P = unknown> = {
  type: "boolean";
  optional?: OptionalFlag<P>;
};
