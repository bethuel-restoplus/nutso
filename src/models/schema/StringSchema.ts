// import { BaseSchema } from "./BaseSchema";
import { ValidationFn } from "../ValidationFn";
import { OptionalFlag } from "../OptionalFlag";

export type StringSchema<R> = {
  type: "string";
  optional?: OptionalFlag<R>;
  minLength?: number;
  maxLength?: number;
  pattern?: RegExp;
  validationFn?: ValidationFn<string, R>;
};
