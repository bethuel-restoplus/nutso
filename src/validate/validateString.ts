import { StringResult } from "../models/result/StringResult";
import { StringSchema } from "../models/schema/StringSchema";
import { FieldPath } from "../models/FieldPath";
import { isNil, isString } from "../utils/typeChecker";

export const validateString = (o: any, schema: StringSchema): StringResult => {
  //

  // isnil
  if (isNil(o)) {
    if (!schema.optional) {
      return {
        isValid: false,
        errorMessage: `Required field.`,
        errorPath: []
      };
    }
    return {
      isValid: true,
      errorMessage: "",
      errorPath: []
    };
  }

  if (!isString(o)) {
    return {
      isValid: false,
      errorMessage: `Should be a string.`,
      errorPath: []
    };
  }

  const str = o as string;

  // min length
  if (!isNil(schema.minLength) && str.length < schema.minLength!) {
    return {
      isValid: false,
      errorMessage: `Should be at least ${schema.minLength} characters.`,
      errorPath: []
    };
  }

  // max length
  if (!isNil(schema.maxLength) && str.length > schema.maxLength!) {
    return {
      isValid: false,
      errorMessage: `Should not be longer than ${schema.maxLength} characters.`,
      errorPath: []
    };
  }

  // pattern
  if (!isNil(schema.regex)) {
    const match = schema.regex!.test(o);
    if (!match) {
      return {
        isValid: false,
        errorMessage: `Should match the pattern ${schema.regex} .`,
        errorPath: []
      };
    }
  }

  return {
    isValid: true,
    errorMessage: ``,
    errorPath: []
  };
};
