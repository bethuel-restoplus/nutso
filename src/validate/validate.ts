import { Schema } from "../models/schema/Schema";
import { Result } from "../models/result/Result";
import { validateString } from "./validateString";
import { StringSchema } from "../models/schema/StringSchema";
import { NumberSchema } from "../models/schema/NumberSchema";
import { validateNumber } from "./validateNumber";
import { BooleanSchema } from "../models/schema/BooleanSchema";
import { validateBoolean } from "./validateBoolean";
import { ObjectSchema } from "../models/schema/ObjectSchema";
import { validateObject } from "./validateObject";
import { FieldPath } from "../models/FieldPath";
import { validateArray } from "./validateArray";
import { ArraySchema } from "../models/schema/ArraySchema";
import { validateDate } from "./validateDate";
import { DateSchema } from "../models/schema/DateSchema";

/**
 * This is an internal API that takes the root as an argument
 * and called recursively from `validateObject` and `validateArray` methods
 *
 * @param o
 * @param root
 * @param schema
 */
export const _validate = <T, R>(o: T | null, root: R, schema: Schema<T, R>): Result<T> => {
  //
  if (!schema) {
    throw new Error("Schema should not be null");
  }

  /**
   * validate object against the schema
   */
  switch (schema.type) {
    case "number":
      return validateNumber(o, root, schema as NumberSchema<R>) as Result<T>;
    case "string":
      return validateString(o, root, schema as StringSchema<R>) as Result<T>;
    case "boolean":
      return validateBoolean(o, root, schema as BooleanSchema<R>) as Result<T>;
    case "array":
      return validateArray(o as any, root, schema as ArraySchema<T, R>) as Result<T>;
    case "object":
      return validateObject(o, root, schema as ObjectSchema<T, R>) as Result<T>;
    case "date":
      return validateDate(o, root, schema as DateSchema<R>) as Result<T>;
  }
  throw new Error(`Unhandled data type`);
};

export const validate = <T>(o: T, schema: Schema<T>): Result<T> => {
  return _validate(o, o, schema);
};
