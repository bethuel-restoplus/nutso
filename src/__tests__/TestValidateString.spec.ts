import { StringSchema } from "../models/schema/StringSchema";
import { validateString } from "../validate/validateString";
import { validate } from "../validate/validate";
import { Schema } from "../models/schema/Schema";

test(`Basic`, () => {
  const str = "foo";
  const schema: StringSchema = {
    type: "string",
  };
  const result1 = validateString({ value: str, schema, parent: undefined as any });
  expect(result1.isValid).toBe(true);

  const result2 = validate(str, schema);
  expect(result2.isValid).toBe(true);
});

test(`Empty string - invalid`, () => {
  const str = "";
  const schema: StringSchema = {
    type: "string",
  };
  const result1 = validateString({ value: str, schema, parent: undefined as any });
  expect(result1.isValid).toBe(false);

  const result2 = validate(str, schema);
  expect(result2.isValid).toBe(false);
});

test(`Basic - invalid min-length`, () => {
  const str = "";
  const schema: StringSchema = {
    type: "string",
    minLength: 3,
  };
  const result1 = validateString({ value: str, schema, parent: undefined as any });
  expect(result1.isValid).toBe(false);

  const result2 = validate(str, schema);
  expect(result2.isValid).toBe(false);
});

test(`Basic - invalid max-length`, () => {
  const str = "foo-bar";
  const schema: StringSchema = {
    type: "string",
    maxLength: 3,
  };
  const result1 = validateString({ value: str, schema, parent: undefined as any });
  expect(result1.isValid).toBe(false);

  const result2 = validate(str, schema);
  expect(result2.isValid).toBe(false);
});

test(`Regex match`, () => {
  const schema: StringSchema = {
    type: "string",
    pattern: /5\d{3}/,
  };
  const valid = "5085";
  const invalid = "508";

  const result1 = validateString({ value: valid, schema, parent: undefined as any });
  expect(result1.isValid).toBe(true);

  const result2 = validate(invalid, schema);
  expect(result2.isValid).toBe(false);
});

test(`Validation function`, () => {
  const schema: StringSchema = {
    type: "string",
    validationFn: (value) => {
      return {
        errorMessage: "Custom validation failed",
      };
    },
  };
  const valid = "5085";

  const result = validateString({ value: valid, schema, parent: undefined as any });
  expect(result).toMatchSnapshot();
});

test(`Validation function - throw error`, () => {
  const schema: StringSchema = {
    type: "string",
    validationFn: (value) => {
      throw new Error(`Validation fn threw!`);
    },
  };
  const valid = "5085";

  const result = validateString({ value: valid, schema, parent: undefined as any });
  expect(result).toMatchSnapshot();
});

test(`Validation function - throw error object - valid`, () => {
  const schema: StringSchema = {
    type: "string",
    validationFn: (value) => {
      throw { message: "Custom error object, with message field" };
    },
  };
  const valid = "5085";

  const result = validateString({ value: valid, schema, parent: undefined as any });
  expect(result).toMatchSnapshot();
});

test(`Validation function - throw error object - invalid`, () => {
  const schema: StringSchema = {
    type: "string",
    validationFn: (value) => {
      throw { foo: "Custom error object, with message field" };
    },
  };
  const valid = "5085";

  const result = validateString({ value: valid, schema, parent: undefined as any });
  expect(result).toMatchSnapshot();
});
