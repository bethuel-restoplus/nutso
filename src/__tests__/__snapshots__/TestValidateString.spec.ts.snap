// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validation function - throw error 1`] = `
Object {
  "errorMessage": "Validation fn threw!",
  "errorPath": Array [],
  "isValid": false,
}
`;

exports[`Validation function - throw error object - invalid 1`] = `
Object {
  "errorMessage": "Exception in validationFn",
  "errorPath": Array [],
  "isValid": false,
}
`;

exports[`Validation function - throw error object - valid 1`] = `
Object {
  "errorMessage": "Custom error object, with message field",
  "errorPath": Array [],
  "isValid": false,
}
`;

exports[`Validation function 1`] = `
Object {
  "errorMessage": "Custom validation failed",
  "errorPath": Array [],
  "isValid": false,
}
`;
